File: processors/key-analyser.c
#define safe_note_count 120
#define safe_buffer_size 32768

double min_peak = 0.0001;

__attribute__((import_module("env")))
__attribute__((import_name("js_log")))
void js_log(double);

__attribute__((used))
double sin_table[safe_note_count] = {0};

__attribute__((used))
double cos_table[safe_note_count] = {0};

__attribute__((used))
double input_buffer[safe_buffer_size] = {0};

__attribute__((used))
double output_buffer_chroma[safe_buffer_size*safe_note_count] = {0};

__attribute__((used))
double output_buffer_peak[safe_buffer_size] = {0};

// __attribute__((used))
// double max_values_sqr[safe_note_count] = {0};

// __attribute__((used))
// unsigned int cutoffs[safe_note_count] = {0};

// __attribute__((used))
// double max_value_sqr = 0;

__attribute__((used))
double sensitivities_sqr[safe_note_count] = {0};

typedef struct {
    double n_1;
    double n_2;
} S;

S sss[safe_buffer_size][safe_note_count] = {{{0}}};

long mm[safe_buffer_size] = {0};
long min_m = 0;
unsigned int m_count = 1;
// this is NOT the index which contains min_m
// it's the index of the max m, but it's the smallest index
unsigned int start_m_index = 0;

double all_time_peak = 0;

double get_value (unsigned int note, S* s) {

	double cos = cos_table[note];
	double sin = sin_table[note];
	double y_r = s->n_1 - cos * s->n_2;
	double y_i = sin * s->n_2;

	return (y_r*y_r + y_i*y_i); // / max_value_sqr;

}

void process_output (unsigned long dft_size, unsigned int note_count, unsigned int m_index, unsigned long output_offset) {

	output_buffer_peak[output_offset] = 1; // initial peak

	unsigned long chroma_offset = output_offset*note_count;

	unsigned int note;

	for (note = 0; note < note_count; note++) {
		output_buffer_chroma[note + chroma_offset] = 0;
	}

	for (note = 0; note < note_count; note++) {

		S* s = &sss[m_index][note];

		double value = get_value(note, s) / (dft_size*dft_size) * 4 * sensitivities_sqr[note];

		output_buffer_chroma[note + chroma_offset] += value;

		s->n_1 = 0;
		s->n_2 = 0;

	}

	double peak = 0;

	for (note = 0; note < note_count; note++) {
		double value = output_buffer_chroma[note + chroma_offset];

		if (value > peak) {
			peak = value;
		}
	}

	all_time_peak = peak / 1024 + all_time_peak * 1023 / 1024;

	if (all_time_peak < min_peak) {
		all_time_peak = min_peak;
	}

	if (all_time_peak > peak) {
		peak = all_time_peak;
	} else {
		all_time_peak = peak / 64 + all_time_peak * 63 / 64;
	}

	if (peak <= 0) return;

	for (note = 0; note < note_count; note++) {
		output_buffer_chroma[note + chroma_offset] /= peak;
	}

	output_buffer_peak[output_offset] = peak;

	return;

}

int process_input (unsigned long dft_size, unsigned long dft_interval, unsigned long note_count, unsigned long buffer_size, int skip_output) {

	unsigned long output_count = 0;

	min_m += buffer_size;

	if (min_m > dft_interval) {

		min_m = min_m - dft_interval;

		mm[start_m_index + m_count] = min_m - buffer_size;

		m_count++;

	}

	unsigned long m_index = start_m_index;

	unsigned long m_i;

	for (m_i = 0; m_i < m_count; m_i++) {

		int m_has_output = 0;

		long m = mm[m_index];

		unsigned long n;

		for (n = 0; n < buffer_size; n++) {

			if (m < 0) {
				m++;
				
				continue;
			}

			double x_n = input_buffer[n];

			unsigned int note;

			// double cutoff = cutoffs[note];

			for (note = 0; note < note_count; note++) {

				S* s = &sss[m_index][note];

				double cos = cos_table[note];

				double s_n = x_n + 2 * cos * s->n_1 - s->n_2;
				s->n_2 = s->n_1;
				s->n_1 = s_n;
				
			}

			m++;
			
			if (m >= dft_size) {

				m_has_output = 1;

				break;

			}

		}

		if (m_has_output > 0) {

			if (!skip_output) {
				process_output(dft_size, note_count, m_index, output_count);
				output_count++;
			}
			
			mm[m_index] = 0;

			start_m_index = (start_m_index + 1) % safe_buffer_size;
			
			m_count--;

		} else {

			mm[m_index] = m;

		}

		m_index = (m_index + 1) % safe_buffer_size;

	}

	return output_count;

}

File: processors/key-analyser.js
class keyAnalyserProcessor extends AudioWorkletProcessor {

	constructor(options) {

		super()

		const processor = this

		const processorOptions = options.processorOptions

		const {
			dftSize,
			dftSampleInterval,
			dftChannels,
			startNote,
			noteCount,
			sampleRate, 
			keyWASM,
			safeNoteCount,
			safeBufferSize
		} = processorOptions

		processorOptions.origDFTSize = dftSize
		processorOptions.origDFTSampleInterval = dftSampleInterval

		processor.options = processorOptions

		const modules = []

		for (let i = 0; i < dftChannels; i++) {
			modules.push(createWASMModule())
		}

		processor._modules = modules

		function createWASMModule() {

			return WebAssembly.instantiate(
				keyWASM,
				{
					env: {
						'js_log': console.log
					}
				}
			).then(wa => {
			
				const sinTable = []
				const cosTable = []
				// const maxValuesSqr = []
				// const cutoffs = []
				// const maxValueSqr = (dftSize / 2) ** 2

				for (let f = 0; f < noteCount; f++) {
					const note = startNote + f
					const frequency = 440 * 2**((note - 69) / 12)
					const cyclesPerSample = frequency / sampleRate
					// const samplesPerCycle = sampleRate / frequency
					// const maxValue = samplesPerCycle / 2
					// const maxValueSqr = maxValue**2

					sinTable[f] = Math.sin((cyclesPerSample % 1) * 2 * Math.PI)
					cosTable[f] = Math.cos((cyclesPerSample % 1) * 2 * Math.PI)
					
					// maxValuesSqr[f] = maxValueSqr

					// for(; m > cutoffSamples; m--) {
					// 	cutoffs[m] = f
					// }

				}

				// for(; m > 0; m--) {
				// 	cutoffs[m] = noteCount
				// }

				const exports = wa.instance.exports
				const buffer = exports.memory.buffer

				;(new Float64Array(buffer, exports.sin_table, safeNoteCount)).set(sinTable)
				;(new Float64Array(buffer, exports.cos_table, safeNoteCount)).set(cosTable)
				// ;(new Float64Array(buffer, exports.max_values_sqr, safeNoteCount)).set(maxValuesSqr)
				// ;(new Uint32Array(buffer, exports.cutoffs, safeNoteCount)).set(cutoffs)
				// ;(new Float64Array(buffer, exports.max_value_sqr, 1)).set([maxValueSqr])

				const inputBuffer = new Float64Array(buffer, exports.input_buffer, safeBufferSize)
				const outputBufferChroma = new Float64Array(buffer, exports.output_buffer_chroma, safeBufferSize * safeNoteCount)
				const outputBufferPeak = new Float64Array(buffer, exports.output_buffer_peak, safeBufferSize)

				return {
					exports,
					inputBuffer,
					outputBufferChroma,
					outputBufferPeak
				}

			}).catch(err => {

				console.error(err)
				console.log('Failed to import WebAssembly key analyser.')
				processor.error = err

				processor.port.postMessage(err)
				
			})

		}

	}

	process(inputs, outputs, parameters) {

		const processor = this

		if (processor.error) return false

		try {

			const modules = processor._modules
			const processorOptions = processor.options
			
			const buffers = inputs.flat()

			const scheduledDate = Date.now()

			Promise.all(modules).then(modules => {
			
				let {
					dftSize,
					dftSampleInterval,
					origDFTSize,
					origDFTSampleInterval,
					dftChannels,
					startNote,
					noteCount,
					sampleRate, 
					keyWASM,
					safeNoteCount,
					safeBufferSize,
					cutoffs
				} = processorOptions

				if (processor.error) return

				const maxDelay = dftSampleInterval / sampleRate * 1000 - 1

				const startDate = Date.now()

				const skipOutput = startDate - scheduledDate > maxDelay

				const outputs = []

				for (let i = 0; i < modules.length; i++) {

					const module = modules[i]
					const buffer = buffers[i]

					const {
						exports,
						inputBuffer,
						outputBufferChroma,
						outputBufferPeak
					} = module

					if (!buffer) continue

					inputBuffer.set(buffer)

					const outputCount = exports.process_input(dftSize, dftSampleInterval, noteCount, Math.min(buffer.length, safeBufferSize), skipOutput)

					if (outputCount > 0) {

						const frames = []

						for (let i = 0; i < outputCount; i++) {
							frames.push([outputBufferChroma.slice(i*noteCount, i*noteCount + noteCount), outputBufferPeak.at(i)])
						}

						outputs.push(frames)
						
					}

				}

				if (outputs.length > 0) {
					this.port.postMessage({outputs})
				}

				if (Date.now() - startDate > maxDelay) {

					console.log('Excess delay detected in key processor, attempting to change settings...')

					if (dftSize > 8192) {

						dftSize = dftSize / 2
						processorOptions.dftSize = dftSize
						console.log('DFT size set to:', dftSize)

					} else if (dftSampleInterval < 32768) {

						dftSize = origDFTSize
						processorOptions.dftSize = dftSize
						console.log('DFT size set to:', dftSize)

						dftSampleInterval = dftSampleInterval * 2
						processorOptions.dftSampleInterval = dftSampleInterval
						console.log('DFT interval set to:', dftSampleInterval)

						this.port.postMessage({
							func: "setKeyInterval",
							args: [dftSampleInterval / sampleRate]
						})

					}

				}

			}).catch(err => {

				console.error(err)
				console.log('Failed to run WebAssembly key analyser.')
				processor.error = err

				processor.port.postMessage(err)
				
			})

			return true

		} catch (err) {

			this.error = err

			throw err

		}

	}

}

registerProcessor('key-analyser', keyAnalyserProcessor)